File: algorithms/sorting/counting-sort/counting_sort.cpp
#include <algorithm>
#include <vector>

std::vector<int> countingSort(const std::vector<int>& numbers) {
    if (numbers.empty()) return {};

    const auto [minimumIterator, maximumIterator] = std::minmax_element(numbers.begin(), numbers.end());
    const int minimum = *minimumIterator;
    const int maximum = *maximumIterator;
    std::vector<std::size_t> counts(static_cast<std::size_t>(maximum - minimum + 1), 0);

    for (const int value : numbers) ++counts[static_cast<std::size_t>(value - minimum)];
    for (std::size_t index = 1; index < counts.size(); ++index) {
        counts[index] += counts[index - 1];
    }

    std::vector<int> output(numbers.size());
    for (auto iterator = numbers.rbegin(); iterator != numbers.rend(); ++iterator) {
        const std::size_t bucket = static_cast<std::size_t>(*iterator - minimum);
        --counts[bucket];
        output[counts[bucket]] = *iterator;
    }
    return output;
}

File: algorithms/sorting/counting-sort/CountingSort.java
import java.util.Arrays;

public final class CountingSort {
    private CountingSort() {
    }

    public static int[] countingSort(int[] numbers) {
        if (numbers.length == 0) return new int[] {};

        int minimum = numbers[0];
        int maximum = numbers[0];
        for (int value : numbers) {
            minimum = Math.min(minimum, value);
            maximum = Math.max(maximum, value);
        }

        int[] counts = new int[maximum - minimum + 1];
        for (int value : numbers) counts[value - minimum]++;
        for (int index = 1; index < counts.length; index++) {
            counts[index] += counts[index - 1];
        }

        int[] output = new int[numbers.length];
        for (int index = numbers.length - 1; index >= 0; index--) {
            int value = numbers[index];
            counts[value - minimum]--;
            output[counts[value - minimum]] = value;
        }
        return output;
    }
}
